objc_msgSend.aarch64.S

```diff
@@ -41,59 +107,128 @@
+	EH_NOP
+	stp    fp, lr, [sp, #192]             // The order is arbitrary, except that
+	EH_SAVE_FP_LR(192)                    // fp and lr must be spilled together
+
+	add    fp, sp, 192					  // Adjust frame pointer
+	EH_ADD_FP(192)
+	stp    \receiver, x8, [sp, #-16]!     // it's convenient if \receiver is spilled at sp
+	EH_STACK_ALLOC(16)					  // stp performed pre-indexing by sp-16
+
+	EH_END_PROLOGUE
+
+	#ifndef _WIN64
 	.cfi_def_cfa fp, 16
 	.cfi_offset fp, -16
 	.cfi_offset lr, -8
+	#endif
 	                                      // We now have all argument registers, the link
 	                                      // register and the receiver spilled on the
 	                                      // stack, with sp containing
 	                                      // the address of the receiver
 
 	mov    x0, sp                         // &self, _cmd in arguments
 	mov    x1, \sel
-	bl     CDECL(slowMsgLookup)           // This is the only place where the CFI directives
+	bl     CDECL(slowMsgLookup)           // This is the only place where the EH directives
 	                                      // have to be accurate...
 	mov    x9, x0                         // IMP -> x9
+
+	EH_START_EPILOGUE
 	ldp    x0, x1, [sp, #16]              // Reload spilled argument registers
+	EH_NOP
 	ldp    x2, x3, [sp, #32]
+	EH_NOP
 	ldp    x4, x5, [sp, #64]
+	EH_NOP
 	ldp    x6, x7, [sp, #64]
+	EH_NOP
 	ldp    q0, q1, [sp, #80]
+	EH_NOP
 	ldp    q2, q3, [sp, #112]
+	EH_NOP
 	ldp    q4, q5, [sp, #144]
+	EH_NOP
 	ldp    q6, q7, [sp, #176]
+	EH_NOP
 	ldp    fp, lr, [sp, #208]
+	EH_SAVE_FP_LR(208)
+
+										  // Post-increment sp += ARGUMENT_SPILL_SIZE +16
 	ldp    \receiver, x8, [sp], #(ARGUMENT_SPILL_SIZE + 16)
+	EH_STACK_ALLOC((ARGUMENT_SPILL_SIZE + 16))
+
+	EH_END_EPILOGUE
+	EH_END_AT_OFFSET(\fnname)
+
 	br     x9
 6:
-	adrp   x10, :got:SmallObjectClasses
-	ldr    x10, [x10, :got_lo12:SmallObjectClasses]
+										  // Load 63:12 of SmallObjectClasses address
+										  // We use the CDECL macro as Windows prefixes
+										  // cdecl conforming symbols with "_".
+	adrp   x10, CDECL(SmallObjectClasses) // The macro handles this transparently.
+
+										  // Add lower 12-bits of SmallObjectClasses address to x10
+	add    x10, x10, :lo12:CDECL(SmallObjectClasses)
 	ldr    x9, [x10, x9, lsl #3]
+
 	b      1b
-	.cfi_endproc
+	EH_END(\fnname)
 .endm
 
+#ifdef _WIN64
+.text
+.def objc_msgSend;
+.scl 2;
+.type 32;
+.endef
+.def objc_msgSend_fpret;
+.scl 2;
+.type 32;
+.endef
+.def objc_msgSend_stret;
+.scl 2;
+.type 32;
+.endef
+#endif
+
 .globl CDECL(objc_msgSend_fpret)
 TYPE_DIRECTIVE(CDECL(objc_msgSend_fpret), %function)
 .globl CDECL(objc_msgSend)
 TYPE_DIRECTIVE(CDECL(objc_msgSend), %function)
 .globl CDECL(objc_msgSend_stret)
 TYPE_DIRECTIVE(CDECL(objc_msgSend_stret), %function)
-CDECL(objc_msgSend):
 CDECL(objc_msgSend_fpret):
 CDECL(objc_msgSend_stret):
-	MSGSEND x0, x1
+CDECL(objc_msgSend):
+	MSGSEND objc_msgSend x0, x1
```

Comment:
It would probably simplify the code to remove the macro here.  On x86, we need a bunch of different variants, but on AArch64 they're all the same.

Reply:
Makes debugging easier too!

Reply:
But what about the function name that changes depending on which objc_msgSend is invoked? We need the name for .seh_startproc.

Reply:
FWIW, I just checked, and it looks like symbol name passed to `.seh_startproc` actually isn't used for anything of value. It looks like you actually could pass any random string as symbol name to it. (WIthin LLVM, it seems like the symbol you pass only is used for printing error messages.)

Also, it's rather relevant here, that when you have a `.seh_proc` that start at a different position than the function itself, the unwind info only starts covering things from the .seh_proc` location. This also means that the 1:1 instruction<->opcode correspondence works out correctly as intended.

Reply:
On all platforms, this function plays somewhat fast and loose with EH mechanisms.  The only thing that *should* be able to cause / propagate exceptions is the call to the slow path, which may end up throwing exceptions out of `+initialize`.  If the lookup succeeds, we tail call the real IMP and so exceptions from that message don't need to be propagated.  In both cases, we don't do any cleanup here and so just need the unwind to work correctly.

As a result, we need the correct unwind table info to be set up on that one branch instruction, we don't care if it's broken everywhere else in the function.

On 32-bit i386 Windows, the setjmp-style mechanism worked very well for us because it unwinds through any function that doesn't directly interact with SEH without running any cleanups.  Everywhere else, we do the absolute minimum to replicate this behaviour.

Reply:
> FWIW, I just checked, and it looks like symbol name passed to `.seh_startproc` actually isn't used for anything of value. It looks like you actually could pass any random string as symbol name to it. (WIthin LLVM, it seems like the symbol you pass only is used for printing error messages.)
> 
> Also, it's rather relevant here, that when you have a `.seh_proc` that start at a different position than the function itself, the unwind info only starts covering things from the .seh_proc` location. This also means that the 1:1 instruction<->opcode correspondence works out correctly as intended.

Alright. I will just pass `objc_msgSend` then, and remove the macro.